state-chain/pallets/cf-ingress-egress/src/boost_pool.rs

```diff
@@ -0,0 +1,338 @@
+
+				(booster_id.clone(), booster_to_receive)
+			})
+			.collect();
+
+		// This shouldn't saturate due to rounding contributions up:
+		let excess_contributed = total_contributed.saturating_sub(required_amount);
+		// This shouldn't saturate due to rounding amounts to receive down:
+		let remaining_to_receive = amount_to_receive.saturating_sub(to_receive_recorded);
+
+		// Some "lucky" booster will receive both of the above (inconsequential) amounts to
+		// ensure that we correctly account for every single atomic unit even in presence
+		// of rounding errors:
+		use nanorand::{Rng, WyRand};
+		let lucky_index = WyRand::new_seed(boost_id).generate_range(0..self.amounts.len());
+		if let Some((lucky_id, amount)) = self.amounts.iter_mut().nth(lucky_index) {
+			amount.saturating_accrue(excess_contributed);
+
+			if let Some(amount) = boosters_to_receive.get_mut(lucky_id) {
+				amount.saturating_accrue(remaining_to_receive)
+			}
+		}
+
+		// For every active booster, record how much of this particular deposit they are owed,
+		// (which is their pool share at the time of boosting):
+		self.pending_boosts
+			.try_insert(boost_id, boosters_to_receive)
+			.map_err(|_| "Pending boost id already exists")?;
+
+		Ok(())
+	}
+
+	pub(crate) fn on_finalised_deposit(
+		&mut self,
+		boost_id: BoostId,
+	) -> Vec<(AccountId, C::ChainAmount)> {
+		let Some(boost_contributions) = self.pending_boosts.remove(&boost_id) else {
+			// The deposit hadn't been boosted
+			return vec![];
+		};
+
+		let mut unlocked_funds = vec![];
+
+		for (booster_id, amount) in boost_contributions {
+			// Depending on whether the booster is withdrawing, add deposits to
+			// their free balance or back to the available boost pool:
+			if let Some(pending_deposits) = self.pending_withdrawals.get_mut(&booster_id) {
+				if !pending_deposits.remove(&boost_id) {
+					log::warn!("Withdrawing booster contributed to boost {boost_id}, but it is not in pending withdrawals");
+				}
+
+				if pending_deposits.is_empty() {
+					self.pending_withdrawals.remove(&booster_id);
+				}
+
+				unlocked_funds.push((booster_id, amount.into_chain_amount()));
+			} else {
+				self.add_funds_inner(booster_id, amount);
+			}
+		}
+
+		unlocked_funds
+	}
+
+	// Returns the number of boosters affected
+	pub fn on_lost_deposit(&mut self, boost_id: BoostId) -> usize {
+		let Some(booster_contributions) = self.pending_boosts.remove(&boost_id) else {
+			log_or_panic!("Failed to find boost record for a lost deposit: {boost_id}");
+			return 0;
+		};
+
+		for booster_id in booster_contributions.keys() {
+			if let Some(pending_deposits) = self.pending_withdrawals.get_mut(booster_id) {
+				if !pending_deposits.remove(&boost_id) {
+					log::warn!("Withdrawing booster contributed to boost {boost_id}, but it is not in pending withdrawals");
+				}
+
+				if pending_deposits.is_empty() {
+					self.pending_withdrawals.remove(booster_id);
+				}
```

Comment:
Tbh, I'm not sure what's being suggested here. Not sure how `BoostPoolFragment` wtih, say, splitting would come into play. Are you talking about `amounts` rather than `pending_withdrawals`?

The two methods here do look quite similar, but we are also talking about 10 lines of code or so, where the body of the loop is mostly different. I could maybe create a function `remove_prewitnessed_deposit` which would return (somewhat unexpectedly) a bool indicating whether the booster is "withdrawing", which we could use to do different things in `on_finalised_deposit` and `on_lost_deposit`, but that doesn't look like a good/clear interface. It seems to me that in this case the (cognitive) cost of indirection isn't worth marginal benefit from deduplicating a few lines of code. We can probably make equally good arguments either way TBH. Maybe I'm misunderstanding your suggestion?